### ISSUE TITLE
Feature Request: Handle "singleton style" content in a better way

### ISSUE THREAD

**Issue by kevmodrome** [pocketbase/pocketbase#159]
If you're building something that needs singleton style data you have to create a collection for each such piece of data and then have one single row in there.  For example, if you're using PocketBase to store marketing text for your landing page, etc.

It would be great if this could be abstracted over in some way. It seems like a common enough use-case that it would be nice.

**Comment by ganigeorgiev** [comment]
I agree that this will be nice to have. I'm just not sure what would be the best way to implement it without complicating the API too much.

I think we can keep the current internals (aka. storing each collection and records in their own table), but provide an option in the Admin UI when you create a new collection to choose between a "Single record" or "Multiple records" collection.

Or in other words, I propose to have only different visual representation with minimal server-side changes, so that when the user pick the "Single record" option, in the Admin UI:
- we will create automatically the collection record in the background (this collection will be allowed to have only one record)
- we will directly show the record values on the collection page, instead of the current records table listing
- we will show only the **view** and **update** API rules in the "Single record" collection settings panel, since create and delete will be not allowed (there is always delete option for the entire collection)

Doing this we would preserve most of the existing internals and in the same time provide a slightly better dev experience.
I'll add it to the roadmap but for now it will be after the import/export feature.

**Comment by kevmodrome** [comment]
I think that makes sense. Good ideas. 👍

**Comment by tryffel** [comment]
Wouldn't a simple key-value Resource solve this problem?

**Comment by ganigeorgiev** [comment]
@tryffel I guess it depends on your use case. You can store mostly everything in a JSON field but you would loose the various filtering and validations that comes with the different field types.

I can think of several other use cases for this feature request:
- for simple reporting pages, internally consisting of a single record with several computed/rawsql fields
- for global app settings/options (eg. app colors, various boolean/toggle controls, etc.)

**Comment by hungcrush** [comment]
Oh.. I'm really need this ...

**Comment by hungcrush** [comment]
Hi @ganigeorgiev I'm working on it, 

> we will create automatically the collection record in the background (this collection will be allowed to have only one record)

Could you give me some suggests for this? I will listen after collection is created, if collection type is singleton I will create record automatically



https://user-images.githubusercontent.com/8098916/219852566-840ba515-1716-4be7-8d91-2a744ce0a32b.jpg

**Comment by jgb-solutions** [comment]
+1

**Comment by upgreidas** [comment]
I needed this as well so I decided to make contribution myself, I'm just wondering what's the best way to create default record for singleton collection? What's the best way to ensure that table migration has completed and you can add new record? Is there event hook for completed database migration? Or maybe migration template should include insertion of default record? @ganigeorgiev

**Comment by ganigeorgiev** [comment]
@upgreidas There are no events at the moment for the database migrations. This feature request has very low priority and may not get implemented. I was initially planning to implement it together with the "View collection" type but didn't like the extra conditional behavior that it required and for now is left aside for later consideration. If you decide to contribute, please read the note in https://github.com/pocketbase/pocketbase#contributing and the check the current active tasks in [roadmap](https://github.com/orgs/pocketbase/projects/2/views/1).

**Comment by calumk** [comment]
+ 1 to say a singleton feature would be brilliant. 

I hope you have some time to look into this possibility.

**Comment by nadam131** [comment]
Can't wait

**Comment by sksar** [comment]
Any updates to this one ?

**Comment by sksar** [comment]
Right now a kind of workaround to use a normal table as a singleton is to set the ID validation to just 1. Autogenerate 1 and set 1 as the validation rule. So only that row will exist and nothing else can exist. Also since the ID is always just 1, use it with normal SDK and the JSVM with the ID hardcoded as 1.

**Comment by KhaledGabr** [comment]
A side effect of this would be that Settings (https://pocketbase.io/docs/api-settings/#list-settings) could be implemented as a singleton .. and would exist under "System" in the dashboard just like _superusers and others .. and we would be able to add new fields to it.

**Comment by ahmetarabaci42** [comment]
I've implemented singleton endpoints for single-record collections.

**New endpoints:**
- `GET /api/singletons/{collection}` - Get the single record
- `PATCH /api/singletons/{collection}` - Update or create the single record

**Implementation:** Added in `examples/base/main.go` as an `OnServe` hook.

**Tested:** Both endpoints work correctly with a `site_settings` collection.

Should I open a PR for this?

**Comment by ganigeorgiev** [comment]
@ahmetarabaci42 Please don't spam and see my reply in https://github.com/pocketbase/pocketbase/issues/159#issuecomment-1603207257.

**Comment by manylittlestorms** [comment]
> I agree that this will be nice to have. I'm just not sure what would be the best way to implement it without complicating the API too much.
> 
> I think we can keep the current internals (aka. storing each collection and records in their own table), but provide an option in the Admin UI when you create a new collection to choose between a "Single record" or "Multiple records" collection.
> 
> Or in other words, I propose to have only different visual representation with minimal server-side changes, so that when the user pick the "Single record" option, in the Admin UI:
> 
>     * we will create automatically the collection record in the background (this collection will be allowed to have only one record)
> 
>     * we will directly show the record values on the collection page, instead of the current records table listing
> 
>     * we will show only the **view** and **update** API rules in the "Single record" collection settings panel, since create and delete will be not allowed (there is always delete option for the entire collection)
> 
> 
> Doing this we would preserve most of the existing internals and in the same time provide a slightly better dev experience. I'll add it to the roadmap but for now it will be somewhere after the import/export feature.

As of Aprile 2026, are there any news about this feature request? It would be very nice to have it.

**Comment by ganigeorgiev** [comment]
@manylittlestorms Nothing has changed here and it remains a very low priority.

All existing feature requests and tasks from the roadmap will be "resolved" (read either implemented or declined) with the ["Stage 2" completion](https://github.com/pocketbase/pocketbase/discussions/7612) but there are **no ETAs**.
